Sources/System/IORing.swift

```diff
@@ -0,0 +1,553 @@
+        }
+        // TODO: actually block here instead of spinning
+    }
+
+}
+
+//TODO: omitting signal mask for now
+//Tell the kernel that we've submitted requests and/or are waiting for completions
+internal func _enter(
+    ringDescriptor: Int32, 
+    numEvents: UInt32,
+    minCompletions: UInt32,
+    flags: UInt32
+) throws -> Int32 {
+  // Ring always needs enter right now;
+    // TODO: support SQPOLL here
+    while true {
+        let ret = io_uring_enter(ringDescriptor, numEvents, minCompletions, flags, nil)
+        // error handling:
+        //     EAGAIN / EINTR (try again),
+        //     EBADF / EBADFD / EOPNOTSUPP / ENXIO
+        //     (failure in ring lifetime management, fatal),
+        //     EINVAL (bad constant flag?, fatal),
+        //     EFAULT (bad address for argument from library, fatal)
+        if ret == -EAGAIN || ret == -EINTR {
+            //TODO: should we wait a bit on AGAIN?
+            continue
+        } else if ret < 0 {
+            fatalError(
+                "fatal error in submitting requests: " + Errno(rawValue: -ret).debugDescription
+            )
+        } else {
+            return ret
+        }
+    }
+}
+
+internal func _submitRequests(ring: inout SQRing, ringDescriptor: Int32) throws {
+    let flushedEvents = _flushQueue(ring: &ring)
+    _ = try _enter(ringDescriptor: ringDescriptor, numEvents: flushedEvents, minCompletions: 0, flags: 0)
+}
+
+internal func _getUnconsumedSubmissionCount(ring: inout SQRing) -> UInt32 {
+    return ring.userTail - ring.kernelHead.pointee.load(ordering: .acquiring)
+}
+
+internal func _getUnconsumedCompletionCount(ring: inout CQRing) -> UInt32 {
+    return ring.kernelTail.pointee.load(ordering: .acquiring) - ring.kernelHead.pointee.load(ordering: .acquiring)
+}
+
+//TODO: pretty sure this is supposed to do more than it does
+internal func _flushQueue(ring: inout SQRing) -> UInt32 {
+    ring.kernelTail.pointee.store(
+        ring.userTail, ordering: .releasing
+    )
+    return _getUnconsumedSubmissionCount(ring: &ring)
+}
+
+@usableFromInline @inline(__always)
+internal func _getSubmissionEntry(ring: inout SQRing, submissionQueueEntries: UnsafeMutableBufferPointer<io_uring_sqe>) -> UnsafeMutablePointer<
+    io_uring_sqe
+>? {
+    let next = ring.userTail &+ 1 //this is expected to wrap
+
+    // FEAT: smp load when SQPOLL in use (not in MVP)
+    let kernelHead = ring.kernelHead.pointee.load(ordering: .acquiring)
+
+    // FEAT: 128-bit event support (not in MVP)
+    if next - kernelHead <= ring.array.count {
+        // let sqe =  &sq->sqes[(sq->sqe_tail & sq->ring_mask) << shift];
+        let sqeIndex = Int(
+            ring.userTail & ring.ringMask
+        )
+
+        let sqe = submissionQueueEntries
+            .baseAddress.unsafelyUnwrapped
+            .advanced(by: sqeIndex)
+
+        ring.userTail = next
+        return sqe
+    }
+    return nil
+}
+
+public struct IORing: @unchecked Sendable, ~Copyable {
```

Comment:
Does this need to be Sendable? I found it surprising that SQRing and CQRing are protected by Mutex. If I was writing a single threaded program, I wouldn't want that overhead.

Reply:
yeah, we decided to change that